src/dataflow-types/src/client/controller/storage.rs

```diff
@@ -605,6 +608,33 @@ where
         Ok(())
     }
 
+    async fn snapshot(
+        &mut self,
+        id: GlobalId,
+        as_of: Self::Timestamp,
+    ) -> Result<Vec<(Row, Diff)>, StorageError> {
+        let as_of = Antichain::from_elem(as_of);
+        let mut snapshot = self.state.persist_handles[&id]
+            .read
+            .snapshot(as_of)
+            .await
+            .expect("cannot read snapshot");
+
+        let mut contents = Vec::new();
+
+        while let Some(updates) = snapshot.next().await {
+            for ((source_data, _pid), _ts, diff) in updates {
+                let row = source_data
+                    .expect("cannot read snapshot")
+                    .0
+                    .expect("cannot read snapshot");
+                contents.push((row, diff));
+            }
+        }
+
+        Ok(contents)
+    }
```

Comment:
Can we get either @danhhz or @petrosagg to take a look at this? In particular, I'm wondering if these `expect`s ought to be `?`s.

Reply:
+1

Reply:
@danhhz I added error types for the two places you weren't sure about. Let me know if they look good. It would be nice to include the since in the `ReadBeforeSince` error, but that would mean plumbing a generic parameter into all uses of `StorageError`, which I didn't think was worth it. Let me know if you think I should add it.